File: ios/HelixBrowser/HelixBrowser/Utils/Prefs.swift
import Foundation
import CoreGraphics

class Prefs {
    static let shared = Prefs()
    private let defaults = UserDefaults.standard

    private init() {}

    var searchEngine: String {
        get { defaults.string(forKey: "search_engine") ?? "google" }
        set { defaults.set(newValue, forKey: "search_engine") }
    }

    var homepage: String {
        get { defaults.string(forKey: "homepage") ?? "https://www.google.com" }
        set { defaults.set(newValue, forKey: "homepage") }
    }

    var isAdBlockEnabled: Bool {
        get { defaults.object(forKey: "ad_block") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "ad_block") }
    }

    var isSaveHistoryEnabled: Bool {
        get { defaults.object(forKey: "save_history") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "save_history") }
    }

    var isDesktopMode: Bool {
        get { defaults.bool(forKey: "desktop_mode") }
        set { defaults.set(newValue, forKey: "desktop_mode") }
    }

    var isBlockTrackersEnabled: Bool {
        get { defaults.object(forKey: "block_trackers") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "block_trackers") }
    }

    var isBlockThirdPartyCookies: Bool {
        get { defaults.object(forKey: "block_third_party_cookies") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "block_third_party_cookies") }
    }

    var isDoNotTrackEnabled: Bool {
        get { defaults.object(forKey: "do_not_track") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "do_not_track") }
    }

    var isHttpsUpgradeEnabled: Bool {
        get { defaults.object(forKey: "https_upgrade") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "https_upgrade") }
    }

    var isBlockFingerprintingEnabled: Bool {
        get { defaults.object(forKey: "block_fingerprinting") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "block_fingerprinting") }
    }

    var isBlockPopupsEnabled: Bool {
        get { defaults.object(forKey: "block_popups") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "block_popups") }
    }

    var isRestoreTabsEnabled: Bool {
        get { defaults.object(forKey: "restore_tabs") as? Bool ?? true }
        set { defaults.set(newValue, forKey: "restore_tabs") }
    }

    var defaultZoom: Int {
        get { defaults.object(forKey: "default_zoom") as? Int ?? 100 }
        set { defaults.set(min(max(newValue, Prefs.minZoom), Prefs.maxZoom), forKey: "default_zoom") }
    }

    /// Mute audio in web content by default. Consumers should apply this to
    /// each WKWebView (e.g. via WKWebViewConfiguration / WKWebView media state)
    /// when a web view is created or this value changes.
    var isMuted: Bool {
        get { defaults.bool(forKey: "is_muted") }
        set { defaults.set(newValue, forKey: "is_muted") }
    }

    // MARK: - Zoom helpers

    /// Supported page-zoom range, expressed as a percentage.
    static let minZoom = 50
    static let maxZoom = 300

    /// `defaultZoom` clamped to the supported range, guarding against any
    /// out-of-range value that may have been persisted previously.
    var clampedDefaultZoom: Int {
        min(max(defaultZoom, Prefs.minZoom), Prefs.maxZoom)
    }

    /// The zoom factor ready to assign to `WKWebView.pageZoom` (iOS 14+).
    /// 100% maps to a scale of 1.0.
    var pageZoomScale: CGFloat {
        CGFloat(clampedDefaultZoom) / 100.0
    }
}

File: windows/HelixBrowser/Utils/Prefs.cs
namespace HelixBrowser;

public class Prefs
{
    private static Prefs? _instance;
    public static Prefs Instance => _instance ??= new Prefs();

    private readonly Windows.Storage.ApplicationDataContainer _settings;

    private Prefs()
    {
        _settings = Windows.Storage.ApplicationData.Current.LocalSettings;
    }

    public string SearchEngine { get => Get("search_engine", "google"); set => Set("search_engine", value); }
    public string Homepage { get => Get("homepage", "https://www.google.com"); set => Set("homepage", value); }
    public bool IsAdBlockEnabled { get => GetBool("ad_block", true); set => Set("ad_block", value); }
    public bool IsSaveHistoryEnabled { get => GetBool("save_history", true); set => Set("save_history", value); }
    public bool IsDesktopMode { get => GetBool("desktop_mode", false); set => Set("desktop_mode", value); }
    public bool IsBlockTrackersEnabled { get => GetBool("block_trackers", true); set => Set("block_trackers", value); }
    public bool IsBlockThirdPartyCookies { get => GetBool("block_third_party_cookies", true); set => Set("block_third_party_cookies", value); }
    public bool IsDoNotTrackEnabled { get => GetBool("do_not_track", true); set => Set("do_not_track", value); }
    public bool IsHttpsUpgradeEnabled { get => GetBool("https_upgrade", true); set => Set("https_upgrade", value); }
    public bool IsBlockFingerprintingEnabled { get => GetBool("block_fingerprinting", true); set => Set("block_fingerprinting", value); }
    public bool IsBlockPopupsEnabled { get => GetBool("block_popups", true); set => Set("block_popups", value); }
    public bool IsRestoreTabsEnabled { get => GetBool("restore_tabs", true); set => Set("restore_tabs", value); }
    public int DefaultZoom { get => GetInt("default_zoom", 100); set => Set("default_zoom", value); }

    private string Get(string key, string defaultValue) =>
        _settings.Values.TryGetValue(key, out var v) ? v as string ?? defaultValue : defaultValue;

    private bool GetBool(string key, bool defaultValue) =>
        _settings.Values.TryGetValue(key, out var v) ? v is bool b ? b : defaultValue : defaultValue;

    private int GetInt(string key, int defaultValue) =>
        _settings.Values.TryGetValue(key, out var v) ? v is int i ? i : defaultValue : defaultValue;

    private void Set(string key, object value) => _settings.Values[key] = value;
}
